KAFKA-4706

sharad-develop opened the pull request:
4706 - Unify StreamsKafkaClient instances

dguy committed:
MINOR: Remove unused constructor param from ProcessorStateManager

Remove applicationId parameter as it is no longer used.

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2385 from dguy/minor-remove-unused-param

rajinisivaram committed:
KAFKA-4363; Documentation for sasl.jaas.config property

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2316 from rajinisivaram/KAFKA-4363

rajinisivaram committed:
KAFKA-4590; SASL/SCRAM system tests

Runs sanity test and one replication test using SASL/SCRAM.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2355 from rajinisivaram/KAFKA-4590

hachikuji committed:
MINOR: Some cleanups and additional testing for KIP-88

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Vahid Hashemian <vahidhashemian@us.ibm.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2383 from hachikuji/minor-cleanup-kip-88

rajinisivaram committed:
KAFKA-4580; Use sasl.jaas.config for some system tests

Switched console_consumer, verifiable_consumer and verifiable_producer to use new sasl.jaas_config property instead of static JAAS configuration file when used with SASL_PLAINTEXT.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2323 from rajinisivaram/KAFKA-4580

(cherry picked from commit 3f6c4f63c9c17424cf717ca76c74554bcf3b2e9a)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

dguy committed:
KAFKA-4588: Wait for topics to be created in QueryableStateIntegrationTest.shouldNotMakeStoreAvailableUntilAllStoresAvailable

After debugging this i can see the times that it fails there is a race between when the topic is actually created/ready on the broker and when the assignment happens. When it fails `StreamPartitionAssignor.assign(..)` gets called with a `Cluster` with no topics. Hence the test hangs as no tasks get assigned. To fix this I added a `waitForTopics` method to `EmbeddedKafkaCluster`. This will wait until the topics have been created.

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Matthias J. Sax, Guozhang Wang

Closes #2371 from dguy/integration-test-fix

(cherry picked from commit 825f225bc5706b16af8ec44ca47ee1452c11e6f3)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

dguy committed:
KAFKA-3452 Follow-up: Refactoring StateStore hierarchies

This is a follow up of https://github.com/apache/kafka/pull/2166 - refactoring the store hierarchies as requested

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2360 from dguy/state-store-refactor

(cherry picked from commit 73b7ae0019d387407375f3865e263225c986a6ce)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

mjsax committed:
MINOR: remove ZK from system tests

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2391 from mjsax/kafka-4060-zk-follow-up-system-tests

ijuma committed:
KAFKA-4591; Create Topic Policy follow-up

1. Added javadoc to public classes
2. Removed `s` from config name for consistency with interface name
3. The policy interface now implements Configurable and AutoCloseable as per the KIP
4. Use `null` instead of `-1` in `RequestMetadata`
5. Perform all broker validation before invoking the policy
6. Add tests

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jason Gustafson <jason@confluent.io>

Closes #2388 from ijuma/create-topic-policy-docs-and-config-name-change

(cherry picked from commit fd6d7bcf335166a524dc9a29a50c96af8f1c1c02)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

guozhangwang committed:
KAFKA-3502; move RocksDB options construction to init()

In RocksDBStore, options / wOptions / fOptions are constructed in the constructor, which needs to be dismissed in the close() call; however in some tests, the generated topology is not initialized at all, and hence the corresponding state stores are supposed to not be able to be closed as well since their `init` function is not called. This could cause the above option objects to be not released.

This is fixed in this patch to move the logic out of constructor and inside `init` functions, so that no RocksDB objects will be created in the constructor only. Also some minor cleanups:

1. In KStreamTestDriver.close(), we lost the logic to close the state stores but only call `flush`; it is now changed back to call both.
2. Moved the forwarding logic from KStreamTestDriver to MockProcessorContext to remove the mutual dependency: these functions should really be in ProcessorContext, not the test driver.

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Damian Guy <damian.guy@gmail.com>, Matthias J. Sax <matthias@confluent.io>, Jason Gustafson <jason@confluent.io>

Closes #2381 from guozhangwang/K3502-pure-virtual-function-unit-tests

(cherry picked from commit 1974e1b0e54abe5fdebd8ff3338df864b7ab60f3)
Signed-off-by: Jason Gustafson <jason@confluent.io>

rajinisivaram committed:
KAFKA-4568; Simplify test code for multiple SASL mechanisms

Remove workaround for testing multiple SASL mechanisms using
sasl.jaas.config and the new support for multiple client
modules within a JVM.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Sriharsha Chintalapani <harsha@hortonworks.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2373 from rajinisivaram/KAFKA-4568

(cherry picked from commit 4c492975a05e4a646f7140c560ed403c48925331)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

dguy committed:
MINOR: Fix javadoc typos in KStream#process

interface for `Processor` in comments incorrectly had `transform` rather than `process`.

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Michael G. Noll, Ismael Juma <ismael@juma.me.uk>

Closes #2396 from dguy/minor-javadoc

(cherry picked from commit ccb183f9fca20a37b9b39761dedcd197e3cf2033)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

dguy committed:
HOTFIX: ChangeLoggingKeyValueStore.name() returns null

This class doesn't need to override this method as it is handled appropriately by the super class

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2397 from dguy/hotfix-npe-state-store

(cherry picked from commit 961ebca57fdd4f8f84df8cf26a835f07ea1718c9)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

mjsax committed:
MINOR: refactor streams system test class hierachy

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Eno Thereska, Guozhang Wang

Closes #2392 from mjsax/minor-system-test-rework

(cherry picked from commit d8a77560c2fa2c209353e3ba2366ad3d4cfdf22c)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

cmccabe committed:
KAFKA-4457; Add BrokerApiVersionsCommand

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>, Apurva Mehta <apurva.1618@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2184 from cmccabe/KAFKA-4457

(cherry picked from commit 4a6f2c6cc0647a08f016a2d712a01ec02630cf87)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

Commit:
KAFKA-4060; Follow-up: Throw exceptions when internal topics to create already exist with unexpected number of partitions

Re-branched the trunk and applied the changes to the new branch to simplify commit log.

Author: Hojjat Jafarpour <hojjat@Hojjat-Jafarpours-MBP.local>

Reviewers: Ismael Juma, Damian Guy, Eno Thereska, Guozhang Wang

Closes #2389 from hjafarpour/KAFKA-4060-Remove-ZkClient-dependency-in-Kafka-Streams-followup-from-trunk

Address Ismael's comments upon merging

(cherry picked from commit 8e2cbae8b12cd71f99f13953b744163f68021323)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

rajinisivaram committed:
KAFKA-4589; SASL/SCRAM documentation

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Gwen Shapira <cshapi@gmail.com>, Sriharsha Chintalapani <harsha@hortonworks.com>, Jun Rao <junrao@gmail.com>

Closes #2369 from rajinisivaram/KAFKA-4589

(cherry picked from commit 666abafcc54f8cab64912355dba4c8ada8e44827)
Signed-off-by: Jun Rao <junrao@gmail.com>

kawamuray committed:
KAFKA-4614; Forcefully unmap mmap of OffsetIndex to prevent long GC pause

Issue: https://issues.apache.org/jira/browse/KAFKA-4614

Fixes the problem that the broker threads suffered by long GC pause.
When GC thread collects mmap objects which were created for index files, it unmaps memory mapping so kernel turns to delete a file physically. This work may transparently read file's metadata from physical disk if it's not available on cache.
This seems to happen typically when we're using G1GC, due to it's strategy to left a garbage for a long time if other objects in the same region are still alive.
See the link for the details.

Author: Yuto Kawamura <kawamuray.dadada@gmail.com>

Reviewers: Apurva Mehta <apurva.1618@gmail.com>, Guozhang Wang <wangguoz@gmail.com>, Ismael Juma <ismael@juma.me.uk>,

Closes #2352 from kawamuray/KAFKA-4614-force-munmap-for-index

(cherry picked from commit 5fc530bc483db145e0cba3b63a57d6d6a7c547f2)
Signed-off-by: Jun Rao <junrao@gmail.com>

vahidhashemian committed:
KAFKA-4547; Avoid incorrect reset of paused partitions to the committed offsets

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Jason Gustafson <jason@confluent.io>

Closes #2341 from vahidhashemian/KAFKA-4547

SandeshKarkera committed:
KAFKA-4432; Added support to supply custom message payloads to perf-producer script.

Current implementation of ProducerPerformance creates static payload. This is not very useful in testing compression or when you want to test with production/custom payloads. So, we decided to add support for providing payload file as an input to producer perf test script.

We made the following changes:
1. Added support to provide a payload file which can have the list of payloads that you actually want to send.
2. Moved payload generation inside the send loop for cases when payload file is provided.

Following are the changes to how the producer-performance is evoked:
1. You must provide "--record-size" or "--payload-file" but not both. This is because, record size cannot be guaranteed when you are using custom events.
  e.g. ./kafka-producer-perf-test.sh --topic test_topic --num-records 100000 --producer-props bootstrap.servers=127.0.0.1:9092 acks=0 buffer.memory=33554432 compression.type=gzip batch.size=10240 linger.ms=10 --throughput -1 --payload-file ./test_payloads --payload-delimiter ,
2. Earlier "--record-size" was a required config, now you must provide exactly one of "--record-size" or "--payload-file". Providing both will result in an error.
3. Support for an additional parameter "--payload-delimiter" has been added which defaults to "\n"

Author: Sandesh K <sandesh.karkera@flipkart.com>

Reviewers: dan norwood <norwood@confluent.io>, Jun Rao <junrao@gmail.com>

Closes #2158 from SandeshKarkera/PerfProducerChanges

(cherry picked from commit a37bf5fffa4a607ded14ec055276d76f19df2d50)
Signed-off-by: Jun Rao <junrao@gmail.com>

xvrl committed:
MINOR: reduce verbosity of cache flushes

This log message tends to be extremely verbose when state stores are being restored

Author: Xavier Léauté <xavier@confluent.io>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2412 from xvrl/reduce-verbosity

(cherry picked from commit 50e035dc341bb595decbac0851354a10c90b0e06)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

shikhar committed:
KAFKA-3209: KIP-66: more single message transforms

Renames `HoistToStruct` SMT to `HoistField`.

Adds the following SMTs:
`ExtractField`
`MaskField`
`RegexRouter`
`ReplaceField`
`SetSchemaMetadata`
`ValueToKey`

Adds HTML doc generation and updates to `connect.html`.

Author: Shikhar Bhushan <shikhar@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2374 from shikhar/more-smt

mjsax committed:
KAFKA-4060 and KAFKA-4476 follow up: remove unnecessary InternalTopicManager branch and fixed one copartitioning bug

ZK removed reveal a bug in `StreamPartitionAssigner` but did not fix it properly. This is a follow up bug fix.

Issue:
 - If topic metadata is missing, `StreamPartitionAssigner` should not create any affected tasks that consume topics with missing metadata.
 - Depending downstream tasks should not be create either.
 - For tasks that are not created, no store changelog topics (if any) should get created
 - For tasks that write output to not-yet existing internal repartitioning topics, those repartitioning topics should not get created

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Damian Guy, Guozhang Wang

Closes #2404 from mjsax/kafka-4060-zk-test-follow-up

(cherry picked from commit 0b99bea590842018e8e97e7fd1c71b1471db4d08)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

xvrl committed:
KAFKA-4672; fix source compatibility for lambda expressions

Variance changes introduced in KIP-100 cause compilation failures with lambda expression in Java 8. To my knowledge this only affects the following method

`KStreams.transform(TransformerSupplier<...>, String...)`

prior to the changes it was possible to write:

`streams.transform(MyTransformer::new)`

where `MyTransformer` extends `Transformer`

After the changes the Java compiler is unable to infer correct return types for the lambda expressions. This change fixed this by reverting to invariant return types for transformer suppliers.

please cherry-pick into 0.10.2.x

Author: Xavier Léauté <xavier@confluent.io>

Reviewers: Ismael Juma, Damian Guy, Guozhang Wang

Closes #2402 from xvrl/lambdas-oh-my

(cherry picked from commit 20e957c9b1f389621bb8ae9e9f5554b66cd92a4f)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

enothereska committed:
MINOR: Pass RecordingLevel to MetricConfig in the broker

This is a KIP-104/105 follow-up. Thanks to ijuma for pointing out.

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2350 from enothereska/minor-broker-level-config

(cherry picked from commit 1eb1e2f60ae006144757fc9fc10ab423c58970bb)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

mjsax committed:
HOTFIX: KAFKA-4060 and KAFKA-4476 follow up

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Damian Guy, Guozhang Wang

Closes #2418 from mjsax/kafka-4060-zk-test-follow-up

(cherry picked from commit dd897bdb2e9cc8790d1e0494fb7867a5cd09ecc6)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

Commit:
KAFKA-4060 Follow-up: update docs accordingly

Updated the docs with changes in KAFKA-4060.

Author: Hojjat Jafarpour <hojjat@Hojjat-Jafarpours-MBP.local>
Author: Hojjat Jafarpour <hojjat@HojjatJpoursMBP.attlocal.net>

Reviewers: Ismael Juma, Matthias J. Sax, Guozhang Wang

Closes #2377 from hjafarpour/KAFKA-4060-docs-update

(cherry picked from commit 62206de4e52dbec32b99baffa0ebf7f5303d7b7f)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

guozhangwang committed:
KAFKA-4633; Always using regex pattern subscription in StreamThread

1. In StreamThread, always use subscribe(Pattern, ..) function in order to avoid sending MetadataRequest with specific topic names and cause brokers to possibly auto-create subscribed topics; the pattern is generated as "topic-1|topic-2..|topic-n".

2. In ConsumerCoordinator, let the leader to refresh its metadata if the generated assignment contains some topics that is not contained in the subscribed topics; also in SubscriptionState, modified the verification for regex subscription to against the regex pattern instead of the matched topics since the returned assignment may contain some topics not yet created when joining the group but existed after the rebalance; also modified some unit tests in `KafkaConsumerTest` to accommodate the above changes.

3. Minor cleanup: changed String[] to List<String> to avoid overloaded functions.

4. Minor cleanup: enforced strong typing in SinkNodeFactory and removed unnecessary unchecked tags.

5. Minor cleanup: augmented unit test error message and fixed a potential transient failure in KafkaStreamTest.

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Damian Guy <damian.guy@gmail.com>, Matthias J. Sax <matthias@confluent.io>, Jason Gustafson <jason@confluent.io>

Closes #2379 from guozhangwang/K4633-regex-pattern

(cherry picked from commit 3400d0c3cc07d5d3136bf9a19142b36cca93f92d)
Signed-off-by: Jason Gustafson <jason@confluent.io>

mjsax committed:
KAFKA-4671: Fix Streams window retention policy

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Damian Guy, Eno Thereska, Guozhang Wang

Closes #2401 from mjsax/kafka-4671-window-retention-policy

(cherry picked from commit 79987590e3e96351ff75ce86718801ec605b2419)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

huxihx committed:
KAFKA-4576; Log segments close to max size break on fetch

`FileChannel.read` may not fill the destination buffer even if there
are enough bytes in the channel to do so. Add a couple of utility
methods that ensure this and use them from all the relevant places.

Author: huxi <huxi@zhenrongbao.com>
Author: amethystic <huxi_2b@hotmail.com>
Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jun Rao <junrao@gmail.com>, Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2304 from amethystic/kafka4576_FileChannel_read

(cherry picked from commit 337f576f5979bf8924c5707b338cf4d3c76a53fd)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

cmccabe committed:
KAFKA-4688; Kafka 0.10.1.1 should be available in system tests

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2424 from cmccabe/KAFKA-4688

(cherry picked from commit 567180605c0a8cf74b8d2cad52e2873c83065a95)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

dpkp committed:
MINOR: Add expected Error Codes to ProduceResponse documentation

This is a documentation-only patch discussed on the mailing list. The
intent is to have these changes propagated to the protocol wiki
(https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol) .

Author: Dana Powers <dana.powers@gmail.com>

Reviewers: Grant Henke <granthenke@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #918 from dpkp/produce_response_errors

(cherry picked from commit 3bb0d3cd888c29a002400253d9ff84ce73ebc03c)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

mjsax committed:
MINOR: rework JavaDoc for windowing related public API

 - also some code refactoring and bug fixes

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Damian Guy, Eno Thereska, Guozhang Wang

Closes #2337 from mjsax/javaDocImprovements4

(cherry picked from commit b938c03b0aa7a0cebdaad0cbb22689cc719c6ebc)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

guozhangwang committed:
KAFKA-3896: Fix KStreamRepartitionJoinTest

The root cause of this issue is that in InternalTopicManager we are creating topics one-at-a-time, and for this test, there are 31 topics to be created, as a result it is possible that the consumer could time out during the assignment in rebalance, and the next leader has to do the same again because of "makeReady" calls are one-at-a-time.

This patch batches the topics into a single create request and also use the StreamsKafkaClient directly to fetch metadata for validating the created topics. Also optimized a bunch of inefficient code in InternalTopicManager and StreamsKafkaClient.

Minor cleanup: make the exception message more informative in integration tests.

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Damian Guy, Matthias J. Sax, Jason Gustafson

Closes #2405 from guozhangwang/K3896-fix-kstream-repartition-join-test

(cherry picked from commit 7837d3e54850fc6ecd16f9bc4dd28a02779572c2)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

cmccabe committed:
KAFKA-4548; Add ClientCompatibilityTest to verify features against older brokers

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2262 from cmccabe/KAFKA-4548

(cherry picked from commit 3df60e7eb8b6fb38cd560708f5b909a9b2e61de2)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

Commit:
KAFKA-4687; Fix InvalidTopicException due to topic creation race condition

We now throw the correct TopicExistsException instead.

Author: Andrew Olson <aolson1@cerner.com>

Reviewers: Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2425 from noslowerdna/KAFKA-4687

(cherry picked from commit 8dfb8afb93f3f1026f868466edb66a4cc63b4455)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

ewencp committed:
KAFKA-4673: Fix thread-safety of Python VerifiableConsumer class

Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Jason Gustafson <jason@confluent.io>

Closes #2427 from ewencp/kafka-4673-verifiable-consumer-thread-safety

(cherry picked from commit 9b8d99b1b866a90b07c8a9a61fe79a75d01ae272)
Signed-off-by: Ewen Cheslack-Postava <me@ewencp.org>

pengwei-li committed:
KAFKA-4229; Controller can't start after several zk expired event

Author: pengwei <pengwei.lihuawei.com>

Reviewers: wangguoz.gmail.com

Author: pengwei-li <pengwei.li@huawei.com>
Author: c00353482 <c00353482@huaweiobz.com>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #2175 from pengwei-li/trunk

(cherry picked from commit 3f0168cd35c1baaa5a032e9a534f28a1bd9221b9)
Signed-off-by: Jun Rao <junrao@gmail.com>

junrao committed:
trivial change to clarify comments in KAFKA-4229

dguy committed:
MINOR: close state store in CachingSessionStoreTest

close the sessions store in `After` to release rocksdb resources.

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2430 from dguy/minor-close-sesion-store

(cherry picked from commit 675881b0f1138c6c09ff4700b17f008ea9053aa4)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

vahidhashemian committed:
MINOR: Update consumer group describe output in the documentation

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2409 from vahidhashemian/doc/update_consumer_group_describe_output

(cherry picked from commit 63cf0e58775b8f0b4abf907866bd7c955500902c)
Signed-off-by: Ewen Cheslack-Postava <me@ewencp.org>

jeyhunkarimov committed:
KAFKA-3835; Streams is creating two ProducerRecords for each send via RecordCollector

Author: Jeyhun Karimov <je.karimov@gmail.com>

Reviewers: Matthias J. Sax, Guozhang Wang

Closes #2417 from jeyhunkarimov/KAFKA-3835

(cherry picked from commit abe19fe693d81ea65a018041f157c58ebecd014d)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

mjsax committed:
HOTIFX: streams system test do not start up correctly

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Guozhang Wang, Damian Guy, Eno Thereska

Closes #2428 from mjsax/hotfixSystemTests

benstopford committed:
KAFKA-4596; Throttled reassignment should support partial JSON file

Fixes a logic error in the Reassignment process which throws an exception
if you don't rebalance all partitions.

Author: Ben Stopford <benstopford@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2399 from benstopford/KAFKA-4596

(cherry picked from commit b8c3b7c964c6167be7f8ad72c96733dfa0d648a2)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

omkreddy committed:
KAFKA-2000; Delete topic should also delete consumer offsets.

Author: Manikumar Reddy O <manikumar.reddy@gmail.com>

Reviewers: Jason Gustafson <jason@confluent.io>, Sriharsha Chintalapani <harsha@hortonworks.com>

Closes #1850 from omkreddy/KAFKA-2700-DELETE

guozhangwang committed:
KAFKA-3502: KStreamTestDriver needs to be closed after the test case

Found a few recently added unit tests did not close KStreamTestDriver after the test itself is closed; this can cause RocksDB virtual function called if the contained topology has some persistent store since they will be initialized but not closed in time.

MINOR fix: found that when closing KStreamTestDriver, we need to first flushing all stores before closing any of them; this is triggered from the `KTableKTableLeftJoin.shouldNotThrowIllegalStateExceptionWhenMultiCacheEvictions`.

MINOR fix: in CachingXXXStore, the `name` field is actually used as the cache's namespace, not really the store name or its corresponding topic name. Fixed it by renaming it to `cacheName` and use `this.name()` elsewhere which will call the underlying store's name.

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Ewen Cheslack-Postava, Eno Thereska, Damian Guy

Closes #2432 from guozhangwang/K3502-kstream-builder-test

(cherry picked from commit 2ca3e59bb6054cc8ecf6926a74859efe3df000ba)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

enothereska committed:
HOTFIX: KIP-104 follow-up: remove duplicated per-processor-node throughput metrics, and also rename some other sensors

This is the HOTFIX PR for any issues detected with KIP-104 until code freeze. Note: do not merge until close to code freeze.

The name changes reflect feedback received while writing the documentation.

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Dan Norwood, Guozhang Wang

Closes #2398 from enothereska/hotfix-streams-metrics

(cherry picked from commit 0971e62e50b94a74aae170fb49d1e532e39b2136)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

ijuma committed:
KAFKA-4698; VerifyError in kafka/client/ClientUtils due to -target:jvm-1.7

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jason Gustafson <jason@confluent.io>

Closes #2434 from ijuma/kafka-4698-target-1.7-verify-error

(cherry picked from commit 23f6951961060ad9a07b42f392470a227269e65a)
Signed-off-by: Jason Gustafson <jason@confluent.io>

ijuma committed:
KAFKA-4597; RecordMetadata should return log append time when appropriate

There is a slight change of behaviour: we now complete the `Future` returned from `send`
before the callbacks are invoked. This seems OK and perhaps a little better as the `Future`
can make progress sooner (as it would typically be blocked on a different thread than the
I/O thread that invokes the callbacks).

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>, Jason Gustafson <jason@confluent.io>

Closes #2318 from ijuma/kafka-4597-record-metadata-log-append-time

(cherry picked from commit 8b3c6c0a1c1bb8f098a7013dbf9c2cb97720e106)
Signed-off-by: Jason Gustafson <jason@confluent.io>

huxihx committed:
KAFKA-4684: add kafka-configs.bat for Windows boxes

Add kafka-configs.bat script for Windows.

Author: huxi <huxi@zhenrongbao.com>

Reviewers: Guozhang Wang, Vahid Hashemian

Closes #2419 from amethystic/kafka4684_offer_kafkaconfigs_script

(cherry picked from commit e9426291aa989ac2624905b844724eeffa608773)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

gwenshap committed:
MINOR: Clarify misleading comment in WordCount example

Author: Gwen Shapira <cshapi@gmail.com>

Reviewers: Matthias J. Sax, Guozhang Wang

Closes #2400 from gwenshap/wordcount-comment

(cherry picked from commit 9a7032720f4600099b130a0bb7b578cedc83e5ff)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

wmarshall484 committed:
MINOR: Fix typo in WordCountProcessorDemo

`bin-kafka-console-producer.sh` should be `bin/kafka-console-producer.sh`.

Author: Will Marshall <wcm214@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2410 from wmarshall484/typo-fix

(cherry picked from commit 51e745ef30c777009c2d1f75965e63a3fe7bb474)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

hachikuji committed:
MINOR: Refactor partition lag metric for cleaner encapsulation

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Jiangjie Qin <becket.qin@gmail.com>, Guozhang Wang <wangguoz@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2416 from hachikuji/refactor-partition-lag-cleanup

(cherry picked from commit 135488352cbe646334a274418d7014a91e92057f)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

cmccabe committed:
KAFKA-4630; Implement RecordTooLargeException when communicating with pre-KIP-74 brokers

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2390 from cmccabe/KAFKA-4630

(cherry picked from commit 5b4e299a46a8ee92e450c54880408713606e4f52)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

ijuma committed:
KAFKA-4699; Invoke producer callbacks before completing the future

This behaviour was changed in 8b3c6c0, but it caused interceptor
test failures (which rely on callbacks) and since we’re so close to
code freeze, it’s better to be conservative.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jason Gustafson <jason@confluent.io>

Closes #2440 from ijuma/kafka-4699-callbacks-invoked-before-future-is-completed

(cherry picked from commit 254e3b77d656a610f19efd1124802e073dfda4b8)
Signed-off-by: Jason Gustafson <jason@confluent.io>

hachikuji committed:
HOTFIX: Consumer offsets not properly loaded on coordinator failover

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2436 from hachikuji/hotfix-offset-deletion

(cherry picked from commit 8301cc608e0e3c334da96b546bf6d943a1704024)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

jonfreedman committed:
MINOR: Update kafka-run-class.bat to handle spaces in classpath

handle existing classpath with spaces
list each lib dependency individually

Author: Jon Freedman <jon.freedman@zoho.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2433 from jonfreedman/trunk

(cherry picked from commit dc4d1bb81b37908725e3bcefd21bade9c2297ebd)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

ijuma committed:
KAFKA-4700: Don't drop security configs in `StreamsKafkaClient`

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Eno Thereska, Damian Guy, Guozhang Wang

Closes #2441 from ijuma/streams-kafka-client-drops-security-configs

(cherry picked from commit 31716590454e1bdd53a65836741f316bf4715c9d)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

vahidhashemian committed:
MINOR: Escape '<' and '>' symbols in quickstart streams code snippet

This was missing from [an earlier PR](https://github.com/apache/kafka/pull/2247) that escaped these symbols in another section of the doc.

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2453 from vahidhashemian/doc/escape_lt_gt_in_streams_code

(cherry picked from commit 0c83eddcda8ed1b483a869f0915815204e3faade)
Signed-off-by: Ewen Cheslack-Postava <me@ewencp.org>

guozhangwang committed:
MINOR: Change logging level for ignored maybeAddMetric from debug to trace

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Eno Thereska <eno.thereska@gmail.com>, Xavier Léauté <xavier@confluent.io>, Jason Gustafson <jason@confluent.io>

Closes #2454 from guozhangwang/KMinor-trace-logging-add-metrics-twice

(cherry picked from commit 1eac3f33f5de3490046203a8e408970da161b84e)
Signed-off-by: Jason Gustafson <jason@confluent.io>

mjsax committed:
MINOR: Update JavaDoc for DSL PAPI-API

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Damian Guy, Guozhang Wang

Closes #2413 from mjsax/javaDocImprovements6

(cherry picked from commit a01b72236926d2bcc52dd7d0378147f8fae8117d)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

ijuma committed:
MINOR: Close create topics policy during shutdown and more tests

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Jason Gustafson <jason@confluent.io>

Closes #2443 from ijuma/close-create-topics-policy-during-shutdown

(cherry picked from commit 2f9e9195abada7c2bcbd41147cb9d4c00f3a712d)
Signed-off-by: Jason Gustafson <jason@confluent.io>

mjsax committed:
MINOR: Add Streams system test for broker backwards compatibility

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Damian Guy, Eno Thereska, Guozhang Wang

Closes #2403 from mjsax/addStreamsClientCompatibilityTest

(cherry picked from commit 89fb02aa81fe67f5a506f579e463f30f3020dee5)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

mjsax committed:
MINOR: Streams API JavaDoc improvements

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Damian Guy, Guozhang Wang

Closes #2437 from mjsax/javaDocImprovements7

(cherry picked from commit 4277645d25e3be0276cb443a36c7ebaa0638034c)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

ijuma committed:
KAFKA-4578; Upgrade notes for 0.10.2.0

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>, Michael G. Noll

Closes #2445 from ijuma/kafka-4578-upgrade-notes-0.10.2

(cherry picked from commit 1956b462527972217447c5e2a5743df8547451c1)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

kamalcph committed:
MINOR: Include more detail in `ConfigDef.parseType` exception message

Author: Kamal C <kamal.chandraprakash@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2345 from Kamal15/trunk

(cherry picked from commit ec76ea343616df27ae2a4e42f5078e081d5878e2)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

cmccabe committed:
KAFKA-4664; Update docs/protocol.html with KIP-97 information

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Gwen Shapira <cshapi@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2387 from cmccabe/KAFKA-4664

(cherry picked from commit 4a14adbb63205e052b7620420b306e29320a4343)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

mjsax committed:
MINOR: Update KTable JavaDoc

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Damian Guy <damian.guy@gmail.com>

Closes #2438 from mjsax/ktableJavaDocs

(cherry picked from commit 1bd11b23ab55b25d5669813f47ad1d5670c6ec9c)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

dguy committed:
KAFKA-4644: Improve test coverage of StreamsPartitionAssignor

Some exception paths not previously covered. Extracted `ensureCopartitioning` into a static class.

Author: Damian Guy <damian.guy@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2448 from dguy/KAFKA-4644

(cherry picked from commit 662c2f893c95128a9d5033f730f1538d368fd151)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

Commit:
MINOR: Replace for within for each; replace if-elseif with match

In some Places for the loop was used but it can be replaced by the for each.
In One file if else if else was used so I replaced the same with match.

Author: Akash Sethi <akash.sethi@knoldus.in>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2435 from akashsethi24/trunk

(cherry picked from commit 9213de8f14dfd2754d913667ae9a703f76277022)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

hachikuji committed:
KAFKA-4704; Coordinator cache loading fails if groupId is reused for offset storage after group is removed

Author: Jason Gustafson <jason@confluent.io>

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #2455 from hachikuji/KAFKA-4704

(cherry picked from commit 57f0cb29973bb40e76ed6deb64b934b4b2e0592a)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

ewencp committed:
MINOR: Update copyright year in the NOTICE file.

Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2460 from ewencp/minor-update-notice-year

(cherry picked from commit 1b964396b88b3d6eea3148fe7c8b4b870910cd8b)
Signed-off-by: Ewen Cheslack-Postava <me@ewencp.org>

rajinisivaram committed:
KAFKA-4557; Handle Producer.send correctly in expiry callbacks

When iterating deque for expiring record batches, delay the
invocation of the callback and deallocation until iteration is
complete since callbacks invoked during expiry may send more
records, modifying the deque, resulting in a
ConcurrentModificationException in the iterator.

Author: Rajini Sivaram <rajinisivaram@googlemail.com>

Reviewers: Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2449 from rajinisivaram/KAFKA-4557

(cherry picked from commit aef6927a42cde193dd6acc36cd4d9f32167da622)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

mjsax committed:
MINOR: update JavaDocs for Kafka Streams DSL helpers

 - also deprecate ZK config for Streams

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Ismael Juma, Guozhang Wang

Closes #2459 from mjsax/javaDocImprovements8

(cherry picked from commit 4c42654b1eecebae272dfe5ce018b85ad4db7709)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

cmccabe committed:
KAFKA-4635; Client Compatibility follow-ups

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #2414 from cmccabe/KAFKA-4635

(cherry picked from commit 8827a5b34ed2a2a570219fb8f29d955918867796)
Signed-off-by: Ismael Juma <ismael@juma.me.uk>

ewencp committed:
KAFKA-4450; Add upgrade tests for 0.10.1 and rename TRUNK to DEV_BRANCH to reduce confusion

Author: Ewen Cheslack-Postava <me@ewencp.org>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #2457 from ewencp/kafka-4450-upgrade-tests

baluchicken committed:
KAFKA-4679: Remove unstable markers from Connect APIs

ewencp ignore this PR if you are already started to work on this ticket.

Author: Balint Molnar <balintmolnar91@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #2423 from baluchicken/KAFKA-4679

mjsax committed:
MINOR: JavaDoc markup cleanup

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2462 from mjsax/javaDocImprovements9

(cherry picked from commit 413444c0ec01ea51019df436cc1f2b664ac61e5a)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

mjsax commented:
This PR need to get rebased to `trunk` before it can get reviewed. Also make sure, that you get the JIRA assigned before starting to work on it, to avoid overlap with any other development effort and thus double work and wasted time.

jeyhunkarimov committed:
KAFKA-4613: Treat null-key records the same way for joins and aggreations

Author: Jeyhun Karimov <je.karimov@gmail.com>

Reviewers: Matthias J. Sax, Eno Thereska, Damian Guy, Guozhang Wang

Closes #2420 from jeyhunkarimov/KAFKA-4613

(cherry picked from commit b559846f2f8151e41a6259e8a570b9422932c836)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

mjsax committed:
KAFKA-4613: Follow-up to fix JavaDocs

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #2468 from mjsax/kafka-4613-null-keys-follow-up

(cherry picked from commit e108a8b4ed4512b021f9326cf079517523c83060)
Signed-off-by: Guozhang Wang <wangguoz@gmail.com>

sharad-develop commented:
Closing pull request. PR not opened as per standards.
